File: src/main/java/com/college/hotlittlepigs/security/AccessManager.java
package com.college.hotlittlepigs.security;

import com.college.hotlittlepigs.user.UserService;
import com.college.hotlittlepigs.user.enums.Role;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.context.annotation.Lazy;
import org.springframework.security.core.context.SecurityContextHolder;
import org.springframework.stereotype.Component;

@Component("accessManager")
public class AccessManager {

  private final UserService userService;

  public AccessManager(@Lazy UserService userService) {
    this.userService = userService;
  }

  public boolean isOwner(Long id) {
    var email = (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();

    var user = userService.getByEmail(email);
    return user.getId().equals(id);
  }

  public boolean isAdmin() {
    var email = (String) SecurityContextHolder.getContext().getAuthentication().getPrincipal();
    var user = userService.getByEmail(email);

    return user.getRole() == Role.ADMIN;
  }
}

File: src/main/kotlin/com/college/hotlittlepigs/security/AccessManager.kt
package com.college.hotlittlepigs.security

import com.college.hotlittlepigs.user.User
import com.college.hotlittlepigs.user.UserService
import com.college.hotlittlepigs.user.enums.Role
import org.springframework.security.core.context.SecurityContextHolder
import org.springframework.stereotype.Component

// TODO add the exception of this component on controller advice
@Component
class AccessManager {
    private lateinit var userService: UserService

    fun isOwner(id: Long): Boolean = currentUser.id == id

    val isAdmin: Boolean
        get() = currentUser.role === Role.ADMIN

    val currentUser: User
        get() = currentUser()

    private fun currentUser(): User {
        val email = SecurityContextHolder.getContext().authentication.principal as String
        return userService.getByEmail(email)
    }


}